### PR TITLE
Multiple works done

### PR DESCRIPTION
**Pull request by chewxy**
* Use of `tensor.Tensor` speeds up FMA 
* Parallelized operations
* Cleaned up `Trainer`

### CODE REVIEW

File: model/word2vec/cbow.go [hunk 1]
```diff
@@ -16,24 +16,37 @@ package word2vec
...
+	// sum, pool tensor.Tensor
+	sums, pools chan tensor.Tensor
+
+	// sum  vector.Vector
+	// pool vector.Vector
 }
 
 // NewCBOW creates *CBOW
 func NewCBOW(s *State) *CBOW {
+	maxprocs := runtime.GOMAXPROCS(-1)
```
Comment (chewxy) [inline model/word2vec/cbow.go]: Should be `runtime.NumCPU()`

File: model/word2vec/skipgram.go [hunk 2]
```diff
@@ -16,22 +16,28 @@ package word2vec
...
 type SkipGram struct {
 	*State
 
-	pool vector.Vector
+	poolvecs chan tensor.Tensor
 }
 
 // NewSkipGram creates *SkipGram
 func NewSkipGram(s *State) *SkipGram {
+	maxprocs := runtime.GOMAXPROCS(-1)
```
Comment (chewxy) [inline model/word2vec/skipgram.go]: should be runtime.NumCPU()

File: model/word2vec/state.go [hunk 3]
```diff
@@ -95,49 +103,55 @@ func (s *State) Preprocess(f io.ReadSeeker) (io.ReadCloser, error) {
...
 	s.startTraining()
 	defer func() {
 		s.endTraining()
 		f.Close()
 	}()
 
+	go s.incrementDoneWord()
+
+	errChan := make(chan error, 1)
+	sema := make(chan struct{}, runtime.GOMAXPROCS(-1))
```
Comment (chewxy) [inline model/word2vec/state.go]: Should be runtime.NumCPU()

File: main.go [hunk 4]
```diff
@@ -15,13 +15,30 @@
...
 	if err := cmd.RootCmd.Execute(); err != nil {
 		os.Exit(1)
 	}
+
+	// f, err := os.Create("mem.prof")
+	// if err != nil {
+	// 	log.Fatal(err)
+	// }
+	// pprof.WriteHeapProfile(f)
+	// f.Close()
```
Comment (ynqa) [inline main.go]: please delete all commented out codes

File: main.go [hunk 5]
```diff
@@ -15,13 +15,30 @@
...
 )
 
 func main() {
+
+	f, err := os.Create("test.prof")
+	if err != nil {
+		log.Fatal(err)
+	}
+	pprof.StartCPUProfile(f)
+	defer pprof.StopCPUProfile()
```
Comment (ynqa) [inline main.go]: I'd like to change these profiling code as optional.
So when only given flag (such as --prof), this code run.

File: model/word2vec/cbow.go [hunk 6]
```diff
@@ -16,24 +16,37 @@ package word2vec
...
 type CBOW struct {
 	*State
 
-	sum  vector.Vector
-	pool vector.Vector
+	// sum, pool tensor.Tensor
+	sums, pools chan tensor.Tensor
+
+	// sum  vector.Vector
+	// pool vector.Vector
```
Comment (ynqa) [inline model/word2vec/cbow.go]: deleting commented out.

File: model/word2vec/skipgram.go [hunk 7]
```diff
@@ -16,22 +16,28 @@ package word2vec
...
-	"github.com/ynqa/word-embedding/vector"
+	"github.com/chewxy/gorgonia/tensor"
 )
 
 // SkipGram is a piece of Word2Vec model.
 type SkipGram struct {
 	*State
 
-	pool vector.Vector
+	poolvecs chan tensor.Tensor
```
Comment (ynqa) [inline model/word2vec/skipgram.go]: It unifies with cbow itself: `pools chan tensor.Tensor`.

File: model/word2vec/state.go [hunk 8]
```diff
@@ -95,49 +103,55 @@ func (s *State) Preprocess(f io.ReadSeeker) (io.ReadCloser, error) {
...
-		words := strings.Fields(line)
-		wordIDs := s.toIDs(words)
-		if err := s.trainOneLine(wordIDs, trainOne); err != nil {
-			return err
+			lower(line)
 		}
 
+		wg.Add(1)
+		wordIDs := s.toIDs(line)
+		go s.trainOneBatch(wordIDs, &wg, sema, errChan, trainOne)
```
Comment (ynqa) [inline model/word2vec/state.go]: [TODO] Next, preparing the following code with the flag like `--debug`:

```go
select {
case err := <- errChan:
        return err
default:
        go s.trainOneBatch(wordIDs, &wg, sema, errChan, trainOne)
}
```

File: model/word2vec/tensor.go [hunk 9]
```diff
@@ -15,17 +15,68 @@
...
+	"math/rand"
+
+	"github.com/chewxy/gorgonia/tensor"
+	"github.com/ynqa/word-embedding/model"
 )
 
-// Tensor has the vector for each word.
-type Tensor []vector.Vector
+var dtype tensor.Dtype = tensor.Float64
+var eng tensor.Engine = tensor.Float64Engine{}
```
Comment (ynqa) [inline model/word2vec/tensor.go]: [ask] Is it better to set `--dtype` flag for given type from outer?

**Comment by ynqa** [comment]
[nits] Please fulfill about golint outputs:
```
model/model.go:31:6: exported type SyncTensor should have comment or be unexported
model/sigmoid.go:23:1: comment on exported function SigmoidF64 should be of the form "SigmoidF64 ..."
model/sigmoid.go:33:1: exported function SigmoidF32 should have comment or be unexported
model/word2vec/tensor.go:35:6: exported type Embedding should have comment or be unexported
```

File: model/word2vec/state.go [hunk 10]
```diff
@@ -95,49 +103,55 @@ func (s *State) Preprocess(f io.ReadSeeker) (io.ReadCloser, error) {
...
 		f.Close()
 	}()
 
+	go s.incrementDoneWord()
+
+	errChan := make(chan error, 1)
+	sema := make(chan struct{}, runtime.NumCPU())
+	var wg sync.WaitGroup
+
 	var current int
```
Comment (ynqa) [inline model/word2vec/state.go]: [nits] I confused this `current` with `currentWordSize` (it's not in this PR target, but I'd like to change this name). What do you think about?

For example:
- current: buffer (chunk)
- currentWordSize: stay this name or `trainedWordSize`

File: model/word2vec/tensor.go (hunk 9 above)
Comment (chewxy) [reply model/word2vec/tensor.go]: yes

File: model/word2vec/state.go (hunk 10 above)
Comment (chewxy) [reply model/word2vec/state.go]: I agree with both changes. I'll put both in

Comment (chewxy) [reply model/word2vec/state.go]: Changes:

`current`: `buffered`
`currentWordSize`: `trainedWords`
`ignoreWord`: `ignoredWords`

**Pull request merged**